src/core/dev_api/openvino/core/dimension_tracker.hpp

```diff
@@ -16,68 +16,43 @@ namespace ov {
     };
 
-    static void set_label(ov::Dimension& d, label_t label) {
-        OPENVINO_ASSERT(label != no_label, "Can not set zero as label for dimension -- it is reserved for no label");
-        d.m_label = label;
-    }
+    static void set_label(ov::Dimension& d, label_t label);
+    static bool has_label(const ov::Dimension& d);
+    static label_t get_label(const ov::Dimension& d);
 
-    static label_t get_label(const ov::Dimension& d) {
-        return d.m_label;
-    }
-
-    void set_table_of_equivalence(ov::Dimension& d) const {
-        OPENVINO_ASSERT(d.m_table_of_equivalence == nullptr, "ov::Dimension is already being tracked");
-        OPENVINO_ASSERT(m_table_of_equivalence != nullptr,
-                        "Can not set nullptr as table of equivalence shared pointer");
-        d.m_table_of_equivalence = m_table_of_equivalence;
-    }
-
-    const std::shared_ptr<TableOfEquivalence>& get_table_of_equivalence(ov::Dimension& d) const {
-        return m_table_of_equivalence;
-    }
-
-    void set_up_for_tracking(ov::Dimension& d, label_t label) const {
-        set_label(d, label);
-        set_table_of_equivalence(d);
-    }
-
-    static void reset_tracking_info(ov::Dimension& d) {
-        d.m_label = no_label;
-        d.m_table_of_equivalence = nullptr;
-    }
+    void set_table_of_equivalence(ov::Dimension& d) const;
+    static const std::shared_ptr<TableOfEquivalence>& get_table_of_equivalence(const ov::Dimension& d);
 
+    void set_up_for_tracking(ov::Dimension& d);
+    void set_up_for_tracking(ov::Dimension& d, label_t label) const;
+    static void reset_tracking_info(ov::Dimension& d);
 private:
     std::shared_ptr<TableOfEquivalence> m_table_of_equivalence;
 };
 
-using EqTable = std::unordered_map<label_t, std::unordered_set<label_t>>;
+using EqualitySoup = std::shared_ptr<std::set<label_t>>;
+using EqTable = std::unordered_map<label_t, EqualitySoup>;
 using ValTable = std::unordered_map<label_t, ov::Dimension>;
 
-class TableOfEquivalence {
+class OPENVINO_API TableOfEquivalence {
 public:
-    void set_as_equal(const ov::Dimension& lhs, const ov::Dimension& rhs) {
-        const auto &l_label = ov::DimensionTracker::get_label(lhs), r_label = ov::DimensionTracker::get_label(rhs);
-        dimension_table_of_equivalence[l_label].insert(r_label);
-        dimension_table_of_equivalence[r_label].insert(l_label);
-    }
-
-    const EqTable& get_equivalence_table() const {
-        return dimension_table_of_equivalence;
-    }
-
-    const ValTable& get_value_equivalence_table() const {
-        return value_table_of_equivalence;
-    }
+    explicit TableOfEquivalence(label_t label = 1) : current_label(label) {};
+    void set_as_equal(const ov::Dimension& lhs, const ov::Dimension& rhs);
+    bool are_equal(const ov::Dimension& lhs, const ov::Dimension& rhs);
 
+    const EqTable& get_equivalence_table() const;
+    const ValTable& get_value_equivalence_table() const;
+    label_t get_next_label();
```

Comment:
Assume this next label is to have unique labels, is that correct?
If this will be unique labels maybe it should be unique for instances of `TableOfEquivalence` or per instance is also OK?

Reply:
It is okay to have unique labels per instance of `TableOfEquivalence` because we have limited scopes of code which label the dimensions and then they clear up all the artifacts of this usage afterwards. So since this is highly controllable usage -- for now this is enough.